src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,317 @@
+        }
+
+        internal void RemoveDisposeScope(DisposeScope disposeScope)
+        {
+            lock (this) {
+                if (DisposeScopeStack.Count == 0) {
+                    throw new InvalidOperationException($"There are no DisposeScopes! ");
+                }
+
+                if (DisposeScopeStack.Peek() != disposeScope) {
+                    throw new InvalidOperationException(
+                        $"DisposeScope for {disposeScope.Disposables} disposed out of order!");
+                }
+
+                DisposeScopeStack.Pop();
+            }
+        }
+
+        internal void MoveToOuterScope(DisposeScope disposeScope, IDisposable disposable)
+        {
+            lock (this) {
+                var array = DisposeScopeStack.ToArray();
+                for (var i = array.Length - 1; i >= 1; i--) {
+                    if (array[i] == disposeScope) {
+                        array[i - 1].Include(disposable);
+                        return;
+                    }
+                }
+            }
+        }
+
+        public class DisposeScope : IDisposable
+        {
+            private readonly DisposeScopeManager _disposeScopeManager;
+
+            public DisposeScope(DisposeScopeManager disposeScopeManager)
+            {
+                _disposeScopeManager = disposeScopeManager;
+                // StackTrace = new StackTrace();
+            }
+
+            // public StackTrace StackTrace { get; set; }
+
+            /// <summary>
+            /// The disposables that are scheduled for disposing.
+            /// </summary>
+            public HashSet<IDisposable> Disposables { get; } = new();
+
+            public long TotalAllocatedSize =>
+                Disposables.OfType<torch.Tensor>().Sum(x => x.NumberOfElements * x.ElementSize);
+
+            public bool DontReportOnExternalDisposes { get; set; } = true;
+
+            /// <summary>
+            /// Includes a disposable in the scope - for tensors this is done automatically once the scope has been
+            /// created. Use this method to add additional disposables that should be disposed, but you typically
+            /// don't need to call this method.
+            /// </summary>
+            /// <param name="disposable">The disposable to keep in the scope</param>
+            /// <returns></returns>
+            public T Include<T>(T disposable) where T : IDisposable
+            {
+                lock (this) {
+                    Disposables.Add(disposable);
+                }
+
+                return disposable;
+            }
+
+            /// <summary>
+            /// Excludes a set of tensors/disposables from the current dispose scope, and moves it up to the outer
+            /// dispose scope, if one exists. See overloaded methods. If you wish to exclude a tensor from all sccopes,
+            /// use ExcludeGlobally.
+            /// </summary>
+            public T Exclude<T>(T exclude) where T : IDisposable
```

Comment:
These overloads will probably cover 90% of all cases. Should there be a public version of the private Exclude() method that allows coverage of the rest?

Reply:
Sounds reasonable. Fixing.